crates/services/txpool_v2/src/service.rs

```diff
@@ -0,0 +1,327 @@
+};
+use parking_lot::RwLock;
+
+use crate::{
+    collision_manager::basic::BasicCollisionManager,
+    config::Config,
+    error::Error,
+    heavy_async_processing::HeavyAsyncProcessor,
+    pool::Pool,
+    ports::{
+        AtomicView,
+        ConsensusParametersProvider,
+        GasPriceProvider as GasPriceProviderTrait,
+        MemoryPool as MemoryPoolTrait,
+        TxPoolPersistentStorage,
+        WasmChecker as WasmCheckerTrait,
+    },
+    selection_algorithms::ratio_tip_gas::RatioTipGasSelection,
+    storage::graph::{
+        GraphConfig,
+        GraphStorage,
+    },
+    verifications::perform_all_verifications,
+};
+
+pub type RemovedTransactions = Vec<PoolTransaction>;
+pub type InsertionResult = Result<RemovedTransactions, Error>;
+
+pub type TxPool<PSProvider> = Arc<
+    RwLock<
+        Pool<
+            PSProvider,
+            GraphStorage,
+            BasicCollisionManager<GraphStorage>,
+            RatioTipGasSelection<GraphStorage>,
+        >,
+    >,
+>;
+
+pub struct SharedState<
+    PSProvider,
+    ConsensusParamsProvider,
+    GasPriceProvider,
+    WasmChecker,
+    MemoryPool,
+> {
+    pool: TxPool<PSProvider>,
+    current_height: Arc<RwLock<BlockHeight>>,
+    consensus_parameters_provider: Arc<ConsensusParamsProvider>,
+    gas_price_provider: Arc<GasPriceProvider>,
+    wasm_checker: Arc<WasmChecker>,
+    memory: Arc<MemoryPool>,
+    heavy_async_processor: Arc<HeavyAsyncProcessor>,
+    utxo_validation: bool,
+}
+
+impl<PSProvider, ConsensusParamsProvider, GasPriceProvider, WasmChecker, MemoryPool> Clone
+    for SharedState<
+        PSProvider,
+        ConsensusParamsProvider,
+        GasPriceProvider,
+        WasmChecker,
+        MemoryPool,
+    >
+{
+    fn clone(&self) -> Self {
+        SharedState {
+            pool: self.pool.clone(),
+            current_height: self.current_height.clone(),
+            consensus_parameters_provider: self.consensus_parameters_provider.clone(),
+            gas_price_provider: self.gas_price_provider.clone(),
+            wasm_checker: self.wasm_checker.clone(),
+            memory: self.memory.clone(),
+            heavy_async_processor: self.heavy_async_processor.clone(),
+            utxo_validation: self.utxo_validation,
+        }
+    }
+}
+
+impl<
+        PSProvider,
+        PSView,
+        ConsensusParamsProvider,
+        GasPriceProvider,
+        WasmChecker,
+        MemoryPool,
+    >
+    SharedState<
+        PSProvider,
+        ConsensusParamsProvider,
+        GasPriceProvider,
+        WasmChecker,
+        MemoryPool,
+    >
+where
+    PSProvider: AtomicView<LatestView = PSView> + 'static,
+    PSView: TxPoolPersistentStorage,
+    ConsensusParamsProvider: ConsensusParametersProvider + 'static,
+    GasPriceProvider: GasPriceProviderTrait + Send + Sync + 'static,
+    WasmChecker: WasmCheckerTrait + Send + Sync + 'static,
+    MemoryPool: MemoryPoolTrait + Send + Sync + 'static,
+{
+    async fn insert(
```

Comment:
We insert transactions into the service, not TxPool=) You can communicate with the service only in a shared state. The internal pipeline of the service may be inserted into the TxPool if it fulfills all requirements.

But I agree that we don't need to have `TxPool` and `HeavyAsyncProcessor` I the shared state. I think we can have a queue of requests that should be processed by the TxPool service, like we did it for P2P.

The `insert` method can return a stream around the `mpsc` channel, where each async task from the `HeavyAsyncProcessor` can submit a transaction when it is inserted or not. In this case, we will have built-in back pressure if the queue is overwhelmed with queries.

In the service `run` loop, we will forward transactions to the `HeavyAsyncProcessor`, which will return the result to the channel after insertion. Or return an error and close the stream(it will fix your TODO).

Reply:
@xgreenx do we want to change this there ?

Reply:
I'm okay to change it in the next PR, and just merge this one to avoid new conflicts

Reply:
Thinking of it, it will means that all read operation such as `find`, `count` etc will need a message passing between shared state and pool instead of a read guard. Are we sure we are not making things worse here ?

Reply:
`find` and `count` shouldn't be used. If we need to use them, then it is okay to fetch this data via the channel. Let's imagine that we have 1000 queries that will cause a `find` operation. All of them will wait until everyone else will use `lock`, plus it will block the TxPool::run function as well. With the channel to pass messages, we have backpressure by default, plus we don't have a lot of simultaneous locks.

Reply:
Okey will implement this in the next pr.